lldb/source/Plugins/DynamicLoader/FreeBSD-Kernel/DynamicLoaderFreeBSDKernel.cpp

```diff
@@ -0,0 +1,770 @@
+  Log *log = GetLog(LLDBLog::DynamicLoader);
+
+  if (addr == LLDB_INVALID_ADDRESS) {
+    if (read_error)
+      *read_error = true;
+    return UUID();
+  }
+
+  LLDB_LOGF(log,
+            "DynamicLoaderFreeBSDKernel::CheckForKernelImageAtAddress: "
+            "looking for kernel binary at 0x%" PRIx64,
+            addr);
+
+  llvm::ELF::Elf32_Ehdr header;
+  if (!ReadELFHeader(process, addr, header))
+    return UUID();
+
+  // Check header type
+  if (header.e_type != llvm::ELF::ET_EXEC)
+    return UUID();
+
+  ModuleSP memory_module_sp =
+      process->ReadModuleFromMemory(FileSpec("temp_freebsd_kernel"), addr);
+  if (!memory_module_sp.get())
+    return UUID();
+
+  ObjectFile *exe_objfile = memory_module_sp->GetObjectFile();
+  if (exe_objfile == nullptr) {
+    LLDB_LOGF(log,
+              "DynamicLoaderFreeBSDKernel::CheckForKernelImageAtAddress "
+              "found a binary at 0x%" PRIx64
+              " but could not create an object file from memory",
+              addr);
+    return UUID();
+  }
+
+  if (is_kernel(memory_module_sp.get())) {
+    ArchSpec kernel_arch(
+        llvm::ELF::convertEMachineToArchName(header.e_machine));
+
+    if (!process->GetTarget().GetArchitecture().IsCompatibleMatch(kernel_arch))
+      process->GetTarget().SetArchitecture(kernel_arch);
+
+    if (log) {
+      std::string uuid_str;
+      if (memory_module_sp->GetUUID().IsValid()) {
+        uuid_str = "with UUID ";
+        uuid_str += memory_module_sp->GetUUID().GetAsString();
+      } else {
+        uuid_str = "and no LC_UUID found in load commands ";
+      }
+      LLDB_LOGF(log,
+                "DynamicLoaderFreeBSDKernel::CheckForKernelImageAtAddress: "
+                "kernel binary image found at 0x%" PRIx64 " with arch '%s' %s",
+                addr, kernel_arch.GetTriple().str().c_str(), uuid_str.c_str());
+    }
+
+    return memory_module_sp->GetUUID();
+  }
+
+  return UUID();
+}
+
+void DynamicLoaderFreeBSDKernel::DebuggerInit(
+    lldb_private::Debugger &debugger) {}
+
+DynamicLoaderFreeBSDKernel::DynamicLoaderFreeBSDKernel(Process *process,
+                                                       addr_t kernel_address)
+    : DynamicLoader(process), m_process(process),
+      m_linker_file_list_struct_addr(LLDB_INVALID_ADDRESS),
+      m_linker_file_head_addr(LLDB_INVALID_ADDRESS),
+      m_kernel_load_address(kernel_address), m_mutex() {
+  process->SetCanRunCode(false);
+}
+
+DynamicLoaderFreeBSDKernel::~DynamicLoaderFreeBSDKernel() { Clear(true); }
+
+void DynamicLoaderFreeBSDKernel::Update() {
+  LoadKernelModules();
+  SetNotificationBreakPoint();
+}
+
+// Create in memory Module at the load address
+bool DynamicLoaderFreeBSDKernel::KModImageInfo::ReadMemoryModule(
+    lldb_private::Process *process) {
+  Log *log = GetLog(LLDBLog::DynamicLoader);
+  if (m_memory_module_sp)
+    return true;
+  if (m_load_address == LLDB_INVALID_ADDRESS)
+    return false;
+
+  FileSpec file_spec(m_name.c_str());
```

Comment:
You can create an `llvm::StringRef` out of a `std::string`. Creating one from a `const char *` involves doing a `strlen`.